{"cells":[{"metadata":{},"cell_type":"markdown","source":"# <font color='blue'> Introduction  </font>\n\nI am editing this copy in order to first of all check whether it works to create viable output and second to track down problems in my adapted code."},{"metadata":{},"cell_type":"markdown","source":"<a id=\"s1\"></a>\n### <font color='blue'> Importing libraries </font>"},{"metadata":{"trusted":true},"cell_type":"code","source":"import numpy as np \nimport pandas as pd \nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\nfrom sklearn import preprocessing                            \nfrom sklearn.decomposition import PCA, FastICA\n\nimport tensorflow as tf\n\nfrom tensorflow import keras\nfrom keras.models import Sequential \nfrom keras import models, utils,backend\nimport keras.utils \nfrom keras.optimizers import Adam   \nfrom keras import layers \nfrom keras.layers import Activation, Dense ,Dropout, BatchNormalization, Input,LeakyReLU\nfrom keras.utils import np_utils\nfrom keras.callbacks import EarlyStopping ,ModelCheckpoint,ReduceLROnPlateau\nfrom sklearn.model_selection import StratifiedKFold\nfrom keras.models import model_from_json  \n\n\nimport sys\nsys.path.append('../input/iterative-stratification/iterative-stratification-master')\nfrom iterstrat.ml_stratifiers import MultilabelStratifiedKFold\n\nimport matplotlib\nimport matplotlib.pyplot as plt\nimport seaborn as sns","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### <font color='blue'> Loading datasets </font>"},{"metadata":{"trusted":true},"cell_type":"code","source":"train = pd.read_csv('/kaggle/input/lish-moa/train_features.csv')\ntest = pd.read_csv('/kaggle/input/lish-moa/test_features.csv')\nsub = pd.read_csv('../input/lish-moa/sample_submission.csv')\ntargets = pd.read_csv('/kaggle/input/lish-moa/train_targets_scored.csv')\ntargets1 = pd.read_csv('/kaggle/input/lish-moa/train_targets_nonscored.csv')\ntargetsnonzero=targets1[targets1.columns[targets1.sum()!=0]]\ntargetsnonzero= targetsnonzero.drop(['sig_id'], axis=1)\ntargets = pd.concat([targets, targetsnonzero], axis=1, sort=False)","execution_count":null,"outputs":[]},{"metadata":{"papermill":{"duration":0.015695,"end_time":"2020-11-02T17:11:23.578847","exception":false,"start_time":"2020-11-02T17:11:23.563152","status":"completed"},"tags":[]},"cell_type":"markdown","source":"<a id=\"s3\"></a>\n## <font color='blue'> Exploring the data</font>"},{"metadata":{},"cell_type":"markdown","source":"We will load two data sets, train and test. Train data set is used for model training and test data set will be the unseen data used to make predictions, lets see now how they look like."},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:23.62128Z","iopub.status.busy":"2020-11-02T17:11:23.620589Z","iopub.status.idle":"2020-11-02T17:11:23.651648Z","shell.execute_reply":"2020-11-02T17:11:23.652129Z"},"papermill":{"duration":0.057067,"end_time":"2020-11-02T17:11:23.652275","exception":false,"start_time":"2020-11-02T17:11:23.595208","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"train.head()","execution_count":null,"outputs":[]},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:23.706656Z","iopub.status.busy":"2020-11-02T17:11:23.695486Z","iopub.status.idle":"2020-11-02T17:11:23.720432Z","shell.execute_reply":"2020-11-02T17:11:23.720952Z"},"papermill":{"duration":0.050434,"end_time":"2020-11-02T17:11:23.721064","exception":false,"start_time":"2020-11-02T17:11:23.67063","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"test.head()","execution_count":null,"outputs":[]},{"metadata":{"papermill":{"duration":0.017644,"end_time":"2020-11-02T17:11:23.80116","exception":false,"start_time":"2020-11-02T17:11:23.783516","status":"completed"},"tags":[]},"cell_type":"markdown","source":"### <font color='blue'>     Missing Values</font>"},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:23.844353Z","iopub.status.busy":"2020-11-02T17:11:23.843439Z","iopub.status.idle":"2020-11-02T17:11:23.920653Z","shell.execute_reply":"2020-11-02T17:11:23.921406Z"},"papermill":{"duration":0.102352,"end_time":"2020-11-02T17:11:23.921605","exception":false,"start_time":"2020-11-02T17:11:23.819253","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"# Lets see size and check for Nulls\nprint('Train dataset',train.shape)\nprint('Test dataset',test.shape)\nmissing_train=(train.isnull().sum()).sum()\nmissing_test=(test.isnull().sum()).sum()\nprint('Missing values in train set:',missing_train,'Missing values in test set:',missing_test)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"categ_feat_train = train.select_dtypes(include=['object'])\nnum_feat_train = train.select_dtypes(exclude=['object'])\nprint('Numerical features',len(num_feat_train.columns))\nprint('Categorical features',len(categ_feat_train.columns))\nprint(categ_feat_train.columns)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"So, there are 876 columns in train and test sets, three of them categorical, and there are no missing values. \n\n* sig_id is the unique primary key of the sample \n* Features with g- : are gene expression levels and there are 772 of them (from g-0 to g-771) \n* Features with c- : are cell viability measurements for each cell line, there are 100 of them (from c-0 to g-99) \n* Features with cp_:  \n  cp_type: samples are treated with a compound(trt_cp) or with a control perturbation (ctl_vehicle) that has no MoAs.   \n  cp_time: duration of the treatment (24, 48 or 72 hours)  \n  cp_dose: dosage of the treatment low/high (D1/D2)\n"},{"metadata":{},"cell_type":"markdown","source":"### <font color='blue'>  Cp_type, cp_time and cp_dose features </font>\nThere are three categorical features cp_type,cp_time and cp_dose. Let us see how their values are distributed in the training dataset. "},{"metadata":{"trusted":true},"cell_type":"code","source":"matplotlib.rc('figure', figsize=(20, 12))\nfig, ax =plt.subplots(2,3)\nsns.countplot(x='cp_type', data=train,ax=ax[0,0])\nax[0,0].set_title('TRAIN- Compound / control treatment', fontsize=14, weight='bold')\nsns.countplot(x='cp_dose', data=train,ax=ax[0,1])\nax[0,1].set_title('TRAIN- Doses  Low / High', fontsize=14,weight='bold')\nsns.countplot(x='cp_time', data=train,ax=ax[0,2])\nax[0,2].set_title('TRAIN- Treatment duration (hours)', fontsize=14,weight='bold')\nsns.countplot(x='cp_type', data=test,ax=ax[1,0])\nax[1,0].set_title('TEST- Compound / control treatment', fontsize=14, weight='bold')\nsns.countplot(x='cp_dose', data=test,ax=ax[1,1])\nax[1,1].set_title('TEST- Doses  Low / High', fontsize=14,weight='bold')\nsns.countplot(x='cp_time', data=test,ax=ax[1,2])\nax[1,2].set_title('TEST- Treatment duration (hours)', fontsize=14,weight='bold')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"As we can see doses and treatment duration times are distributed equally but only 8% of the samples are treated with a control perturbation. Distributions are similar both in training and test sets ,probably indicating that they follow same experiment setup (since each combo of drug-timing-dose can be viewed as one independent experiment, so there are 6 independent realizations per drug, although some drugs have been profiled more than once). "},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> Gene features  </font>\n \nThe role of genes is to encode proteins who dictate how a cell functions. So, genes expressed in a particular cell determine what that cell can do. We can see that genes expression values in our data sets show normal like distribution, with zero mean as random following plots show."},{"metadata":{"trusted":true},"cell_type":"code","source":"matplotlib.rc('figure', figsize=(20, 4))\nfig, ax =plt.subplots(1,4)\nfig.suptitle('Genes distributions', fontsize=16)\nsel_genes = [7,16,33,66]\ni=0\nfor item in sel_genes:\n    train.hist(column=['g-'+ str(item)], ax=ax[i])\n    i+=1\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"matplotlib.rc('figure', figsize=(20, 4))\nfig.suptitle('Genes distributions G-0 to G-771', fontsize=16)\nsel_genes = list(range(0, 771))\nfor item in sel_genes:\n    sns.kdeplot(data=train['g-'+ str(item)], shade=False,legend=False)\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'>Genes correlation analysis </font>\n\nIt would probably be useful to examine if there is any pairwise correlation (suggesting  a biological relationship) between genes, such that changes in the expression levels of one gene correspond to changes in the expression level of another gene.\nThere are references though like [this ](https://www.frontiersin.org/articles/10.3389/fmicb.2015.00650/full) , suggesting that “ …. computation of pairwise gene associations (correlation; mutual information) produces unexpectedly large variation in estimates of pairwise gene association—regardless of the metric used, the organism under study, or the number and source of the samples probably due to sampling bias.” And also “….many individual genes show small differences in absolute gene expression levels across the set of samples. These small differences are due mainly to “noise” instead of “signal” attributable to environmental or genetic perturbations. \nSo we will not examine corellations but we can keep that ICA as we shall see later could be used as a helpful denoising step even though it might be sightly biased towards highly expressed genes."},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> Cell features analysis </font>\n\nThese features are related to cell viability, high negative cell viability values reflect a high numbers of cell deaths and low values high survival rates. Their distributions are skewed with a normal-like look but with heavy negative tails and peaks at -10 probably due to experiment data normalization procedure. "},{"metadata":{"trusted":true},"cell_type":"code","source":"matplotlib.rc('figure', figsize=(20, 4))\nfig, ax =plt.subplots(1,4)\nfig.suptitle('Cells distributions', fontsize=16)\nsel_cells = [7,16,30,45]\ni=0\nfor item in sel_cells:\n    train.hist(column=['c-'+ str(item)], ax=ax[i])\n    i+=1\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"matplotlib.rc('figure', figsize=(20, 4))\nfig.suptitle('Cells distributions c-0 to c-771', fontsize=16)\nsel_cells = list(range(0, 99))\nfor item in sel_cells:\n    sns.kdeplot(data=train['c-'+ str(item)], shade=False,legend=False)\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### <font color='blue'> Cell correlation analysis </font>\nLet's explore how cells are correlated."},{"metadata":{"trusted":true},"cell_type":"code","source":"Cells = [c for c in train.columns if \"c-\" in c]\nplt.figure(figsize=(10,6))\nsns.heatmap(train[Cells].corr(), cmap='viridis')\nplt.title('Cell viability correlations (Train set)', fontsize=14, weight='bold')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"There is a clear high correlation between cell viabilities that has to be examined. "},{"metadata":{},"cell_type":"markdown","source":"<a id=\"s4\"></a>\n## <font color='blue'> Targets </font>"},{"metadata":{},"cell_type":"markdown","source":"There are two target files containing MOAs of interest, the first one (train_targets_scored.csv) contains features that are scored and an auxiliary one (train_targets_nonscored.csv) having elements that are not scored . Let's see what type of features they have"},{"metadata":{"trusted":true},"cell_type":"code","source":"targets.head()","execution_count":null,"outputs":[]},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:24.322818Z","iopub.status.busy":"2020-11-02T17:11:24.321872Z","iopub.status.idle":"2020-11-02T17:11:24.324761Z","shell.execute_reply":"2020-11-02T17:11:24.32524Z"},"papermill":{"duration":0.028614,"end_time":"2020-11-02T17:11:24.325375","exception":false,"start_time":"2020-11-02T17:11:24.296761","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"targets.columns","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The targets we have to predict are basically probabilities of activation for each of various proteins-targets. This is a multi-label problem since one sample can be classified to multiple targets or none. The main target types as we can see are activators, inhibitors, receptors agonists and antagonists, agents, stimulants. \n\n* Receptors are chemical structures, composed of protein, that receive and transduce signals that may be integrated into biological systems.[1] These signals are typically chemical messengers which bind to a receptor and cause some form of cellular/tissue response, e.g. a change in the electrical activity of a cell.\n * Agonists are chemicals that bind to a receptors and activate them to produce a biological response.  \n * An antagonist blocks the action of the agonist, while an inverse agonist causes an action opposite to that of the agonist.\n* Activators : They are proteins that increase transcription of a gene or set of genes. Activators are considered to have positive control over gene expression, as they function to promote gene transcription and, in some cases, are required for the transcription of genes to occur. Most activators are DNA-binding proteins that bind to enhancers  \n* Inhibitors  : An enzyme inhibitor is a molecule that binds to an enzyme and decreases its activity\n   \n  \n\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"target_classes1 = targets.drop(['sig_id'], axis=1).astype(bool).sum(axis=1).reset_index()\ntarget_classes1.columns = ['Sig_Ids', 'activations']\ntarget_classes1 = target_classes1.groupby(['activations'])['Sig_Ids'].count().reset_index()\ntarget_classes2 = targets1.drop(['sig_id'], axis=1).astype(bool).sum(axis=1).reset_index()\ntarget_classes2.columns = ['Sig_Ids', 'activations']\ntarget_classes2 = target_classes2.groupby(['activations'])['Sig_Ids'].count().reset_index()\nmatplotlib.rc('figure', figsize=(8, 5))\nfig, ax =plt.subplots(1,2)\nsns.barplot(x=\"activations\", y=\"Sig_Ids\", data=target_classes1, ax=ax[0]).set_title('Scored targets')\nsns.barplot(x=\"activations\", y=\"Sig_Ids\", data=target_classes2, ax=ax[1]).set_title('NonScored targets')\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We can see that numbers of activations differ for scored and non_scored datasets, but in any case more than 90% of the drugs activate zero or one of the target columns."},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> Removing uninformative features </font>\n\nThe fewer and more useful features we have, the better for our models. An interesting idea could be to try to find and remove any \"uninformative\" genes in our data sets like those ones with very low expression values in most samples or those whose expression values shows small variation throughout samples. Removing those feature could help our models perform better and faster. We have to remember though that what we see comes after z-score and data quantile normalization so we have to do some \"reverse enginineering\", but there is no way to get back to values before transformations since statistics like mean and std are lost. So we will use data as is and try VarianceThreshold to detect and remove low variance features (we have already scaled our data). Target variability was set to 90% with trial and error. "},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.feature_selection import VarianceThreshold\n\ncombo = pd.concat([train, test], axis=0)\ncols=combo.columns.tolist()\ncatego = combo.iloc[:, 0:4]\nthr = VarianceThreshold(0.95)  \n#VTgenes_df = pd.DataFrame(VT_genes, index = combo.index)\n#cells_df = combo[cols[776:]]\nVT_ALL = thr.fit_transform(combo[cols[4:]])\n\nVT_ALL[:5,:15]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"VT_ALL.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"vt_df = pd.DataFrame(VT_ALL)\nranges = pd.Series(vt_df.max()-vt_df.min())\nprint(ranges.mean(),ranges.min(),ranges.max())","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'>  Quantile Transformation </font>\n\nIt might be helpful for our models to change distributions using QuantileTransformer, that provides non-linear transformations in which distances between marginal outliers and inliers are shrunk. StandardScaler and MinMaxScaler were also tested but since they are very sensitive to outliers, I didn't use them."},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.preprocessing import QuantileTransformer\n\nQUA = QuantileTransformer(n_quantiles=100, output_distribution='normal')\n#1 cells_df = QUA.fit_transform(cells_df)\n#1 VTgenes_df = QUA.fit_transform(VTgenes_df)\nVT_ALL = QUA.fit_transform(VT_ALL)\n\n#from sklearn.preprocessing import MinMaxScaler\n#MMscale=MinMaxScaler()\n#VTgenes_df=MMscale.fit_transform(VTgenes_df)\n#cells_df=MMscale.fit_transform(cells_df)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# <font color='blue'> Dimensionality reduction </font>"},{"metadata":{},"cell_type":"markdown","source":"## Using ICA to reduce dimensionality\n\nI examined both PCA and ICA to perform an initial reduction in the dimensionality of the input dataset while still preserving most of the important data structure. There is work presenting the use of ICA in dimensionality reduction, deconvolution, data pre-processing, meta-analysis, and others applied to different data types (transcriptome, methylome, proteome, single-cell data) like [this one](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6771121/) and also references that ICA can be selected over PCA depending on the case. My best results were with ICA so I will use it in this kernel. Determining the optimal number of independent components is a difficult task and is selected by trials, [this article ](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-4112-9) can be very informative.\n<img src=\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6771121/bin/ijms-20-04414-g001.jpg\" width=\"600\">\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"'''\nc_ica = FastICA(n_components=20,max_iter=500)\ncells_ica=c_ica.fit_transform(cells_df)\ng_ica = FastICA(n_components=80,max_iter=500)\ngenes_ica=g_ica.fit_transform(VTgenes_df)\n\nica_c_df = pd.DataFrame(cells_ica , columns=[\"c_PCA\" + str(i) for i in range(20)], index=combo.index)\nica_g_df = pd.DataFrame(genes_ica, columns=[\"g_PCA\" + str(i) for i in range(80)], index=combo.index)\n\nnew_combo= pd.concat([catego,combo[cols[4:]],ica_c_df,ica_g_df], axis=1 )\ntrain = new_combo[ : train.shape[0]]\ntest = new_combo[-test.shape[0] : ]\n'''","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"all_ica = FastICA(n_components=300,max_iter=500)\nVT_ALL_ica=all_ica.fit_transform(VT_ALL)\n\nVT_ALL_ica_df = pd.DataFrame(VT_ALL_ica , columns=[\"VTICA\" + str(i) for i in range(300)], index=combo.index)\nfull= combo.iloc[:, 1:]\n\n\nnew_combo= pd.concat([full,VT_ALL_ica_df], axis=1 )\ntrain = new_combo[ : train.shape[0]]\ntest = new_combo[-test.shape[0] : ]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# <font color='blue'>  Categorical Data </font>\n In the following part we shall convert all categorical features into dummy/indicator variables"},{"metadata":{"trusted":true},"cell_type":"code","source":"train['cp_time'] = train['cp_time'].map( {24: 1, 48: 2, 72: 3} ).astype(int)\ntrain = pd.get_dummies(train, columns = [\"cp_type\"], prefix=\"CPTP\",drop_first=True)\ntrain = pd.get_dummies(train, columns = [\"cp_dose\"], prefix=\"CPD\", drop_first=True)\n\ntest['cp_time'] = test['cp_time'].map( {24: 1, 48: 2, 72: 3} ).astype(int)\ntest = pd.get_dummies(test, columns = [\"cp_type\"], prefix=\"CPTP\",drop_first=True)\ntest = pd.get_dummies(test, columns = [\"cp_dose\"], prefix=\"CPD\",drop_first=True)","execution_count":null,"outputs":[]},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:31.531648Z","iopub.status.busy":"2020-11-02T17:11:31.53038Z","iopub.status.idle":"2020-11-02T17:11:31.725885Z","shell.execute_reply":"2020-11-02T17:11:31.725347Z"},"papermill":{"duration":0.223467,"end_time":"2020-11-02T17:11:31.726003","exception":false,"start_time":"2020-11-02T17:11:31.502536","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"#X = train.drop(['sig_id'], axis=1).values\n#Xtest = test.drop(['sig_id'], axis=1).values\ny = targets.drop(['sig_id'], axis=1)\nX = train.values\nXtest = test.values\n","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> Multi-label Model </font>"},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:37.68593Z","iopub.status.busy":"2020-11-02T17:11:37.683999Z","iopub.status.idle":"2020-11-02T17:11:37.686607Z","shell.execute_reply":"2020-11-02T17:11:37.68706Z"},"papermill":{"duration":0.039684,"end_time":"2020-11-02T17:11:37.687171","exception":false,"start_time":"2020-11-02T17:11:37.647487","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"def build_model():\n    keras.backend.clear_session()\n    model = models.Sequential()\n    n_cols = X.shape[1]\n    model.add(layers.BatchNormalization())\n    model.add(layers.Dropout(0.5))\n    model.add(layers.Dense(2048,activation='relu', input_shape=(n_cols,)))\n    model.add(layers.BatchNormalization())\n    model.add(layers.Dropout(0.5))\n    model.add(layers.Dense(2048,activation='relu' ))\n    model.add(layers.BatchNormalization())\n    model.add(layers.Dropout(0.5))\n    model.add(layers.Dense(537, activation='sigmoid'))\n    return model          ","execution_count":null,"outputs":[]},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:11:37.75734Z","iopub.status.busy":"2020-11-02T17:11:37.749331Z","iopub.status.idle":"2020-11-02T17:34:12.477656Z","shell.execute_reply":"2020-11-02T17:34:12.477081Z"},"papermill":{"duration":1354.768139,"end_time":"2020-11-02T17:34:12.477777","exception":false,"start_time":"2020-11-02T17:11:37.709638","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"reduce_lr = keras.callbacks.ReduceLROnPlateau(monitor='val_logloss', factor=0.3, patience=5, mode='min', min_lr=1E-5)\nearly_stopping = keras.callbacks.EarlyStopping(monitor='val_logloss', min_delta=1E-5, patience=15, mode='min',restore_best_weights=True)\n\nn_labels = y.shape[1]\nn_features = X.shape[1]\nn_train = X.shape[0]\nn_test = Xtest.shape[0]\n\n\n# Label smoothing\n\n# Since our score is heavily punished by very confident (very close to 0/1) incorrect answers\n#as @Rahul suggests, labels will be smoothed to a small extent, and predictions are clipped. There\n# is a build in option for label smoothing in BinaryCrossentropy set to a small value\n\ndef logloss(y_true, y_pred):\n    y_pred = tf.clip_by_value(y_pred,0.001,0.999)\n    return -backend.mean(y_true*backend.log(y_pred) + (1-y_true)*backend.log(1-y_pred))\n\nn_seeds = 6\nnp.random.seed(1)\nseeds = np.random.randint(0,100,size=n_seeds)\n\nn_folds = 5\ny_pred = np.zeros((n_test,n_labels))\noof = tf.constant(0.0)\nhists = []\nfor seed in seeds:\n    fold = 0\n    mltsplit = MultilabelStratifiedKFold(n_splits=n_folds,shuffle=True,random_state=seed)\n    for train, test in mltsplit.split(X,y.values):\n        X_train = X[train]\n        X_test  = X[test]\n        y_train = y.values[train]\n        y_test  = y.values[test]\n\n        model=build_model()\n        model.compile(optimizer=keras.optimizers.Adam(lr=1e-3, decay=1e-3 / 200), loss=tf.keras.losses.BinaryCrossentropy(label_smoothing=0.001), metrics=logloss)\n\n        hist = model.fit(X_train,y_train, batch_size=128, epochs=50,verbose=0,validation_data = (X_test,y_test),callbacks=[reduce_lr, early_stopping])\n        hists.append(hist)\n        model.save('LabelSmoothed_seed_'+str(seed)+'_fold_'+str(fold))\n        y_val = model.predict(X_test)\n        oof += logloss(tf.constant(y_test,dtype=tf.float32),tf.constant(y_val,dtype=tf.float32))/(n_folds*n_seeds)\n        y_pred += model.predict(Xtest)/(n_folds*n_seeds)\n        print('seed=',seed,'fold=',fold)\n        fold += 1\nprint('OOF score is ',oof)\n\nplt.figure(figsize=(12,8))\nhist_train = np.zeros(40)\nhist_val = np.zeros(40)\nfor i in range(n_folds):\n    hist_train += np.array(hists[i].history['loss'][:40])/n_folds\n    hist_val += np.array(hists[i].history['val_loss'][:40])/n_folds\n\nplt.plot(hist_train)\nplt.plot(hist_val)\nplt.yscale('log')\nplt.yticks(ticks=[1,1E-1,1E-2])\nplt.xlabel('Epochs')\nplt.ylabel('Average Logloss')\nplt.legend(['Training','Validation'])","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> Experimenting with model parameters and pre/post processing data </font>\n\nI tested various techniques and parameters during this work, some of them improved my LB significally.  \n\n1) Using cp_type column : predictions after removing this feature were less accurate.  \n2) Using Standard /Minmax scaling but Normal Quantile Transform before ICA lead to better predictions  \n3) Label smoothing is definitely leading to improved scores due to the way of LB scoring system. Trying to use values larger than 0.001/0.999 for min/max lowered scores.  \n4) Adding weight normalization was tested too but without any success  \n5) Testing AdamW, LazyAdam optimizers, didn't improve results  \n6) Testing Leaky_relu ,elu no obvious improvement  \n7) Changing batch size from 128 up to 512 didn't change LB score (as expected)\n8) Using larger models : more hidden layers and/or nodes per layer up to 8192/2048 showed small improvement (possibly due to some extra overfitting)  \n9) Tried batch normalization before the activation function, not suitable for our activations \n10) Tested different values for FastICA independent components"},{"metadata":{"execution":{"iopub.execute_input":"2020-11-02T17:34:12.551095Z","iopub.status.busy":"2020-11-02T17:34:12.550408Z","iopub.status.idle":"2020-11-02T17:34:12.554939Z","shell.execute_reply":"2020-11-02T17:34:12.555673Z"},"papermill":{"duration":0.048094,"end_time":"2020-11-02T17:34:12.555843","exception":false,"start_time":"2020-11-02T17:34:12.507749","status":"completed"},"tags":[],"trusted":true},"cell_type":"code","source":"target_cols= targets.columns[1:207]\nN_TARGETS = len(target_cols)\nprint(N_TARGETS)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"test_preds = sub.copy()\ntest_preds[target_cols] = 0\ntest_preds.loc[:,target_cols] += y_pred[:,:206]\ntest2 = pd.read_csv('/kaggle/input/lish-moa/test_features.csv')\ntest2 = test2.drop(['sig_id'], axis=1)\ntest2 = pd.get_dummies(test2, columns = [\"cp_type\"], prefix=\"CPTP\",drop_first=True)\ntest_preds.iloc[test2['CPTP_trt_cp'] == 0,1:] = 0\ntest_preds.to_csv('submission.csv', index=False)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## <font color='blue'> REFERENCES </font>  \nThere are really many inspiring notebooks and discussions published for this contest covering various aspects, that helped me a lot in to build this kernel to mention a few:\n\n* https://www.kaggle.com/c/lish-moa/discussion/184005\n* https://www.kaggle.com/isaienkov/mechanisms-of-action-moa-prediction-eda\n* https://www.kaggle.com/rahulsd91/moa-label-smoothing\n* https://www.kaggle.com/nayuts/moa-pytorch-nn-pca-rankgauss\n\nThank you very much for your time reading this kernel. Please feel free to leave your comments and suggestions about how I can improve this work. And don't forget , if you found something that you liked or gave you an idea, do UPVOTE!\n        \n    "}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}